### PR TITLE
Add support for using project dependencies in the declarative DSL with DependencyCollector

### PR DESCRIPTION
**Pull request by tresat**
This is pairing work done with @eskatos when attempting to convert additional subprojects in Now In Android.

It is necessary for https://github.com/gradle/nowinandroid/pull/5.

### CODE REVIEW

**Review by big-guy — APPROVED** [review]

File: platforms/core-configuration/declarative-dsl-provider/src/main/kotlin/org/gradle/internal/declarativedsl/project/dependencyConfigurationSchema.kt [hunk 1]
```diff
@@ -148,14 +157,29 @@ class ImplicitDependencyCollectorFunctionResolver(configurations: DependencyConf
...
+                    return RuntimeFunctionResolver.Resolution.Resolved(object : DeclarativeRuntimeFunction {
+                        override fun callBy(receiver: Any, binding: Map<DataParameter, Any?>, hasLambda: Boolean): DeclarativeRuntimeFunction.InvocationResult {
+                            val dependencyNotation = binding.values.single() as ProjectDependency
+                            val collector: DependencyCollector = getterFunction.call(receiver) as DependencyCollector
+                            collector.add(dependencyNotation)
+                            return DeclarativeRuntimeFunction.InvocationResult(Unit, null)
+                        }
+                    })
+                } else {
+                    throw IllegalStateException()
```
Comment (h0tk3y) [inline platforms/core-configuration/declarative-dsl-provider/src/main/kotlin/org/gradle/internal/declarativedsl/project/dependencyConfigurationSchema.kt:181]: Consider adding an exception message here?

**Review by h0tk3y — COMMENTED** [review]
👍overall, see some ideas of improvements.

File: platforms/core-configuration/declarative-dsl-provider/src/main/kotlin/org/gradle/internal/declarativedsl/project/dependencyConfigurationSchema.kt [hunk 2]
```diff
@@ -148,14 +157,29 @@ class ImplicitDependencyCollectorFunctionResolver(configurations: DependencyConf
...
-                        val dependencyNotation = binding.values.single().toString()
-                        val collector: DependencyCollector = getterFunction.call(receiver) as DependencyCollector
-                        collector.add(dependencyNotation)
-                        return DeclarativeRuntimeFunction.InvocationResult(Unit, null)
-                    }
-                })
+                val gavParam = DataParameter("dependency", String::class.toDataTypeRef(), false, ParameterSemantics.Unknown)
+                val projectParam = DataParameter("dependency", ProjectDependency::class.toDataTypeRef(), false, ParameterSemantics.Unknown)
+                if (parameterValueBinding.bindingMap.containsKey(gavParam)) {
+                    return RuntimeFunctionResolver.Resolution.Resolved(object : DeclarativeRuntimeFunction {
```
[the remark this answers is not in the document]
Comment (tresat) [reply platforms/core-configuration/declarative-dsl-provider/src/main/kotlin/org/gradle/internal/declarativedsl/project/dependencyConfigurationSchema.kt:163]: I don't think so.  The problem is the `collector.add(dependencyNotation)` won't work with `Any`, it needs to accept either a `ProjectDependency` or `CharSequence`/`String`, so you end up with a reusable function like:

```
private
    fun invokeAddDependency(
        getterFunction: KFunction<*>,
        receiver: Any,
        dependencyNotation: Any
    ): DeclarativeRuntimeFunction.InvocationResult {
        val collector: DependencyCollector = getterFunction.call(receiver) as DependencyCollector
        when(dependencyNotation) {
            is ProjectDependency -> collector.add(dependencyNotation)
            is String -> collector.add(dependencyNotation)
            else -> error("Unexpected dependency notation: $dependencyNotation")
        }
        return DeclarativeRuntimeFunction.InvocationResult(Unit, null)
    }
```

Which is longer and arguably more complex than the original.

**Comment by tresat** [comment]
@bot-gradle test and merge

**Pull request merged**